apps/rig_tests/test/proxy/publish_to_event_stream/kafka_test.exs

```diff
@@ -0,0 +1,204 @@
+    HTTPoison.post!(rig_api_url, body, headers)
+
+    # The client calls the proxy endpoint:
+    request_url = rig_proxy_url <> endpoint_path
+
+    %HTTPoison.Response{status_code: res_status, body: res_body, headers: res_headers} =
+      HTTPoison.options!(request_url, headers)
+
+    assert res_status == 204
+    assert res_body == ""
+
+    # CORS should verify origin
+    assert Enum.member?(res_headers, {"access-control-allow-origin", "*"})
+    # CORS should verify HTTP methods
+    assert Enum.member?(res_headers, {"access-control-allow-methods", "*"})
+    # CORS should verify allowed headers
+    assert Enum.member?(
+             res_headers,
+             {"access-control-allow-headers", "content-type,authorization"}
+           )
+  end
+
+  @tag :kafka
+  test "Given target is set to Kafka, the http request should publish message to Kafka topic" do
+    test_name = "proxy-publish-to-kafka"
+
+    api_id = "mock-#{test_name}-api"
+    endpoint_id = "mock-#{test_name}-endpoint"
+    endpoint_path = "/#{endpoint_id}"
+
+    # We register the endpoint with the proxy:
+    rig_api_url = "http://localhost:#{@api_port}/v1/apis"
+    rig_proxy_url = "http://localhost:#{@proxy_port}"
+
+    body =
+      Jason.encode!(%{
+        id: api_id,
+        name: "Mock API",
+        version_data: %{
+          default: %{
+            endpoints: [
+              %{
+                id: endpoint_id,
+                type: "http",
+                not_secured: true,
+                method: "POST",
+                path: endpoint_path,
+                target: "kafka"
+              }
+            ]
+          }
+        },
+        proxy: %{
+          target_url: "http://localhost",
+          port: 3000
+        }
+      })
+
+    headers = [{"content-type", "application/json"}]
+    HTTPoison.post!(rig_api_url, body, headers)
+
+    # The client calls the proxy endpoint:
+    request_url = rig_proxy_url <> endpoint_path
+
+    kafka_body =
+      Jason.encode!(%{
+        "data" => %{
+          "eventID" => "069711bf-3946-4661-984f-c667657b8d85",
+          "eventType" => "com.example.test",
+          "eventTypeVersion" => "1.0",
+          "eventTime" => "2018-04-05T17:31:00Z",
+          "cloudEventsVersion" => "0.1",
+          "source" => "/rig-test",
+          "contentType" => "application/json",
+          "extensions" => %{},
+          "data" => %{
+            "foo" => "bar"
+          }
+        },
+        "partition_key" => "test_key"
+      })
+
+    %HTTPoison.Response{status_code: res_status, body: res_body} =
+      HTTPoison.post!(request_url, kafka_body, headers)
+
+    assert res_status == 202
+    assert res_body == "Accepted."
+
+    expected_msg =
+      Jason.encode!(%{
+        "source" => "rig-test",
+        "rig" => %{
+          "scheme" => "http",
+          "requestPath" => "/mock-proxy-http-response-publish-to-kafka-endpoint",
+          "reqHeaders" => [
+            ["user-agent", "hackney/1.14.0"],
```

Comment:
I suppose this makes the test unnecessary brittle. The assertion should cover the `requestPath`, the `correlationID` and the `eventID`, and perhaps also the `method`. If `host` and `port` should be checked as well, then you should probably use `Confex` to fetch the actual values of `HOST`, and `@proxy_port` for the actual port, instead of assuming `localhost` and `4000`.

Reply:
I would rather check everything and be sure when something changes - not a fan of partial checking.

Reply:
But do you really want to go fix tests every time hackney is upgraded? You got me curious.. why are you not a fan of partial checking?

Reply:
That's true, on the other hand if someone extends the message in code and do mistake there, but forget to update test we have error in app without knowing it.

Reply:
I feel the simple question that needs to be asked for each value we test is: is there a reason, in the context of this test, for testing it? Given this example, the test assures that `Given target is set to Kafka, the http request should publish message to Kafka topic`, which describes the following assertion: the message sent to the proxy ends up at the Kafka topic. That's what this is about. Now, how can we check that the message ends up there? We may check the event ID or the whole message, depending on whether we have reason to believe the message could be altered in the process (e.g., the proxy doesn't use the body verbatim, so that's a valid concern). But that's about it - indeed, checking the added meta data could justify a second test, something like "When forwarding an event to Kafka, connection related meta data is added to the given CloudEvent.". In that test, of course, we'd assert the presence and correctness of the connection related meta data, but even in this test I doubt that the user-agent is of any relevance.

Reply:
Just noticed the test is anyway badly written, will do update.